<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CU04_Optimización de vacunas</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 10. Datos meteorológicos de estaciones

* Descarga de datos desde AEMET




<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Encoding

Con la siguiente expresión se evitan problemas con el encoding al ejecutar el notebook. Es posible que deba ser eliminada o adaptada a la máquina en la que se ejecute el código.

In [ ]:
Sys.setlocale(category = "LC_ALL", locale = "es_ES.UTF-8")

[1] "es_ES.UTF-8/es_ES.UTF-8/es_ES.UTF-8/C/es_ES.UTF-8/C"

### Packages to use

- {tcltk} para selección interactiva de archivos locales
- {sf} para trabajar con georeferenciación
- {readr} para leer y escribir archivos csv
- {dplyr} para explorar datos
- {stringr} para manipulación de strings
- {climaemet} para obtener los datos de aemet


In [1]:
library(sf)
library(readr)
library(dplyr)
library(stringr)
library(climaemet)



Linking to GEOS 3.10.2, GDAL 3.4.2, PROJ 8.2.1; sf_use_s2() is TRUE


Attaching package: 'dplyr'


The following objects are masked from 'package:stats':

    filter, lag


The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union




### Paths

In [2]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

No aplica

## Open Data

### Datos de AEMET


1. Obtener API key de AEMET
2. Descargar datos directamente con el paquete `climaemet`
3. Transformar para guardar en CSV

In [3]:
# browseURL("https://opendata.aemet.es/centrodedescargas/obtencionAPIKey")


In [4]:
aemet_api_key("eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJlbWlsaW8ubG9wZXpAdXJqYy5lcyIsImp0aSI6ImI4ZmEyNjMyLTI3ODktNDcwNi1hMGM1LWNmNjI1MWU1NmFlYiIsImlzcyI6IkFFTUVUIiwiaWF0IjoxNjc1NDI4MjEyLCJ1c2VySWQiOiJiOGZhMjYzMi0yNzg5LTQ3MDYtYTBjNS1jZjYyNTFlNTZhZWIiLCJyb2xlIjoiIn0.uf2TizyDvDXQ4G12POBoNFZirdGtT1EW09emo5mvtj8")



To install your API key for use in future sessions, run this function with `install = TRUE`.



**Estaciones que usaremos**

In [5]:
estacionesm <- aemet_stations() |> 
  filter(provincia == "MADRID") |> 
  pull(indicativo)


Datos horarios:

In [6]:
data <-
  aemet_daily_clim(estacionesm, 
                   start = "2021-09-01", 
                   end = "2023-01-31",
                   return_sf = TRUE)


Warning message in is.na(tbl[[lat]]) || any(is.na(tbl[[lon]])):
"'length(x) = 6143 > 1' in coercion to 'logical(1)'"


Estructura de los datos:

In [7]:
glimpse(data)

Rows: 6,143
Columns: 21
$ fecha       <date> 2021-09-01, 2021-09-02, 2021-09-03, 2021-09-04, 2021-09-0~
$ indicativo  <chr> "2462", "2462", "2462", "2462", "2462", "2462", "2462", "2~
$ nombre      <chr> "PUERTO DE NAVACERRADA", "PUERTO DE NAVACERRADA", "PUERTO ~
$ provincia   <chr> "MADRID", "MADRID", "MADRID", "MADRID", "MADRID", "MADRID"~
$ altitud     <dbl> 1894, 1894, 1894, 1894, 1894, 1894, 1894, 1894, 1894, 1894~
$ tmed        <dbl> 13.2, 11.8, 13.8, 16.7, 18.0, 19.4, 17.1, 13.8, 12.6, 12.6~
$ prec        <chr> "2,8", "3,2", "0,0", "0,0", "Ip", "0,0", "0,0", "Ip", "0,3~
$ tmin        <dbl> 8.2, 8.5, 9.4, 11.4, 13.1, 16.3, 13.9, 10.3, 9.5, 8.8, 9.9~
$ horatmin    <chr> "01:00", "02:00", "04:50", "05:20", "03:30", "02:30", "23:~
$ tmax        <dbl> 18.3, 15.2, 18.3, 22.0, 22.9, 22.5, 20.3, 17.3, 15.6, 16.5~
$ horatmax    <chr> "11:50", "14:40", "16:10", "13:20", "12:00", "13:20", "14:~
$ dir         <chr> "16", "24", "21", "19", "17", "15", "14", "21", "27", "32"~
$ velmedia    <d

Muestra primeros datos

In [8]:
data |> tibble() |> slice_head(n = 5)

fecha      indicativo nombre                provincia altitud tmed prec tmin
1 2021-09-01 2462       PUERTO DE NAVACERRADA MADRID    1894    13.2 2,8   8.2
2 2021-09-02 2462       PUERTO DE NAVACERRADA MADRID    1894    11.8 3,2   8.5
3 2021-09-03 2462       PUERTO DE NAVACERRADA MADRID    1894    13.8 0,0   9.4
4 2021-09-04 2462       PUERTO DE NAVACERRADA MADRID    1894    16.7 0,0  11.4
5 2021-09-05 2462       PUERTO DE NAVACERRADA MADRID    1894    18.0 Ip   13.1
  horatmin tmax ... dir velmedia racha horaracha sol  presMax horaPresMax
1 01:00    18.3 ... 16  3.1      15.6  00:30      5.1 816.6   01         
2 02:00    15.2 ... 24  2.2       8.3  19:50      2.9 816.4   23         
3 04:50    18.3 ... 21  2.2       6.7  06:00      7.5 816.4   11         
4 05:20    22.0 ... 19  3.3       8.3  09:10     12.1 816.3   Varias     
5 03:30    22.9 ... 17  3.6      11.1  23:40     11.7 819.5   24         
  presMin horaPresMin geometry                  
1 813.4   04          POINT (-4.010556 40.79306)
2 813.5   03          POINT (-4.010556 40.79306)
3 815.1   06          POINT (-4.010556 40.79306)
4 814.7   05          POINT (-4.010556 40.79306)
5 815.9   04          POINT (-4.010556 40.79306)

## ETL Processes

#### Transform data

* Asignar precipitación 0 a los valores inapreciables
* Convertir la precipitación en numérica

* Convertir objeto espacial en coordenadas numéricas para exportar a csv

In [9]:
tdata <- data |> 
    mutate(prec = str_replace(prec, "Ip", "0"),
           prec = as.numeric(str_replace(prec, ",", "."))) |>
    bind_cols(st_coordinates(data)) |>
    st_drop_geometry()

In [10]:
glimpse(tdata)

Rows: 6,143
Columns: 22
$ fecha       <date> 2021-09-01, 2021-09-02, 2021-09-03, 2021-09-04, 2021-09-0~
$ indicativo  <chr> "2462", "2462", "2462", "2462", "2462", "2462", "2462", "2~
$ nombre      <chr> "PUERTO DE NAVACERRADA", "PUERTO DE NAVACERRADA", "PUERTO ~
$ provincia   <chr> "MADRID", "MADRID", "MADRID", "MADRID", "MADRID", "MADRID"~
$ altitud     <dbl> 1894, 1894, 1894, 1894, 1894, 1894, 1894, 1894, 1894, 1894~
$ tmed        <dbl> 13.2, 11.8, 13.8, 16.7, 18.0, 19.4, 17.1, 13.8, 12.6, 12.6~
$ prec        <dbl> 2.8, 3.2, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.3, 0.0, 0.0, 0.0~
$ tmin        <dbl> 8.2, 8.5, 9.4, 11.4, 13.1, 16.3, 13.9, 10.3, 9.5, 8.8, 9.9~
$ horatmin    <chr> "01:00", "02:00", "04:50", "05:20", "03:30", "02:30", "23:~
$ tmax        <dbl> 18.3, 15.2, 18.3, 22.0, 22.9, 22.5, 20.3, 17.3, 15.6, 16.5~
$ horatmax    <chr> "11:50", "14:40", "16:10", "13:20", "12:00", "13:20", "14:~
$ dir         <chr> "16", "24", "21", "19", "17", "15", "14", "21", "27", "32"~
$ velmedia    <d

## Synthetic Data Generation

No aplica

## Fake Data Generation

No aplica

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar la extensión añadida del fichero para diferenciarlas

<font color='tomato'> Identificamos los datos a guardar</font>

In [11]:
data_to_save <- tdata


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [12]:
caso <- "CU_04"
proceso <- '_05'
tarea <- "_10"
archivo <- ""
proper <- "_aemet"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [13]:
# file_save <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out <- paste0(oPath, file_save)
# write_csv(data_to_save, path_out)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/. 
- Pero en este caso lo guardamos solo en Input porque es una descarga directa y no viene de otro input 

In [14]:
file_save <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out <- paste0(iPath, file_save)
write_csv(data_to_save, path_out)

cat('File saved as: ')
path_out

File saved as: 

[1] "Data/Input/CU_04_05_10_aemet.csv"

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

This working code needs the following conditions:

* For using the interactive selection of file, the {tcltk} package must be installed. It is not needed in production.
* The {xxx} package must be installed. 
* The data paths `Data/Input` and `Data/Output` must exist (relative to the notebook path)

### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* climaemet 1.0.2
* dplyr 1.0.10
* readr 2.1.3
* stringr 1.5.0
* sf 1.0.9


### Data structures

#### Objeto `tdata`

- Los datos de origen se descargan en formato `sf`
- Hay 6143 filas con las variables:
    - fecha
	- indicativo
	- nombre
	- provincia
	- altitud
	- tmed
	- prec
	- tmin
	- horatmin
	- tmax
	- horatmax
	- dir
	- velmedia
	- racha
	- horaracha
	- sol
	- presMax
	- horaPresMax
	- presMin
	- horaPresMin
	- X
	- Y


#### Observaciones generales sobre los datos

* Se han obtenido datos desde septiembre de 2019 hasta enero de 2023. 
* Hay varias variables meteorológicas, suele haber bastantes valores perdidos

### Consideraciones para despliegue en piloto

* Se ha utilizado una api key particular: conseguir una específica para proyecto antes de ejecutar

### Consideraciones para despliegue en producción

* Se deben crear los procesos ETL en producción necesarios para que los datos de entrada estén actualizados
* Se ha utilizado una api key particular: conseguir una específica para proyecto antes de ejecutar

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se han descargado los datos de AEMET por estaciones

#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se debe interpolar la temperatura y precipitaciones a las zonas básicas de salud

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [15]:
# incluir código